analyzers/src/SonarAnalyzer.Common/Rules/InsteadOfAnyBase.cs

```diff
@@ -0,0 +1,124 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class InsteadOfAnyBase<TSyntaxKind, TInvocationExpression> : SonarDiagnosticAnalyzer
+    where TSyntaxKind : struct
+    where TInvocationExpression : SyntaxNode
+{
+    private const string ExistsDiagnosticId = "S6605"; // Collection-specific "Exists" method should be used instead of the "Any" extension.
+    private const string ContainsDiagnosticId = "S6617"; // Collection-specific "Contains" method should be used instead of the "Any" extension.
+    private const string MessageFormat = "Collection-specific \"{0}\" method should be used instead of the \"Any\" extension.";
+
+    private readonly DiagnosticDescriptor existsRule;
+    private readonly DiagnosticDescriptor containsRule;
+    private ImmutableArray<KnownType> ExistsTypes { get; } = ImmutableArray.Create(
```

Comment:
Why is this an instance property instead of static field? You could rewrite them as:

```csharp
    private static readonly ImmutableArray<KnownType> ExistsTypes = ImmutableArray.Create(
        KnownType.System_Array,
        KnownType.System_Collections_Immutable_ImmutableList_T);
```